packages/virtual/src/index.js

```diff
@@ -4,6 +4,7 @@ import path from 'path';
 const PREFIX = `\0virtual:`;
```

Comment:
I see we are doing all this because virtual ids do have this prefix. I know it has been there before but I think it should generally only be a last resort change to update `input`. The reason is that in the future, `input` might be extended in unexpected ways, which means that this plugin needs to play catch-up and will become more brittle.

Moreover I think the  issue is that IMO we should NOT add this prefix. Adding a `\0` prefix signifies to other plugins that they should skip this plugin e.g. in the `transform` hook. This means, however, that neither e.g. `rollup-plugin-typescript` nor `rollup-plugin-babel` will touch these modules, which I do not think is what users will expect. The only issue I foresee at the moment is that another plugin might catch the `load` hook, but I am not sure which plugin that could be. To prevent this, users should be instructed that this plugin should ideally be placed before other plugins that load modules.

What I would recommend instead (and yes, this will be a breaking change, but worthwhile IMO):
- Do not update `input` in the options hook.
- Change the `resolveId` hook like this:
  ```js
  resolveId(id, importer) {
      if (modules.hasOwnProperty(id)) {
        return id;
      }

      if (importer) {
        const resolved = path.resolve(path.dirname(importer), id);
        if (resolvedIds.has(resolved)) return resolved;
      }
  }
  ```
- Change the load hook like this:
  ```js
  load(id) {
      if (modules.hasOwnProperty(id)) {
        return modules[id];
      }
      if (resolvedIds.has(id)) {
        return resolvedIds.get(id);
      }
  }
  ```

Opinions?

Another approach might be to use a prefix without a leading `\0` but as I said, I am not 100% happy about modifying `input`.

Reply:
This is actually then insight I was looking for. WRT prefixes - won't we need something to differentiate virtual modules from static/file-based modules to prevent collisions? Or should that be left up to the user?

Reply:
The only situation I could think of is if a plugin implements a custom load hook. But most that I am aware of are not problematic as they just load themselves plugin-specific virtual files that have their own prefix. The prefixes have always been there (https://github.com/rollup/rollup-plugin-virtual/commit/6d7f6d08dd2803e9d103d26c13561080521f70d0#diff-1fdf421c05c1140f6d71444ea2b27638R3) and I guess they have not been added to solve a particular problem but I could be wrong. Or do you remember something I miss here @Rich-Harris (as it was your initial commit)?